graphql_client_codegen/src/query.rs

```diff
@@ -622,17 +647,20 @@ impl UsedTypes {
     }
 }
 
-fn resolve_variables(
+fn resolve_variables<'doc, T>(
     query: &mut Query,
-    variables: &[graphql_parser::query::VariableDefinition],
+    variables: &[graphql_parser::query::VariableDefinition<'doc, T>],
     schema: &Schema,
     operation_id: OperationId,
-) {
+)
+where
+    T: graphql_parser::query::Text<'doc>,
+{
     for var in variables {
         query.variables.push(ResolvedVariable {
             operation_id,
-            name: var.name.clone(),
-            default: var.default_value.clone(),
+            name: var.name.as_ref().into(),
+            default: var.default_value.as_ref().map(|dflt| dflt.into_static()),
```

Comment:
Needs https://github.com/graphql-rust/graphql-parser/pull/57

Reply:
If we start using `'static` things all over when will those values be freed? Does it happen when the corresponding `Value` gets freed? Using `'static` always strikes me as a little sketchy, but it may certainly be worth it here as long as it doesn't leak memory.

Reply:
It's `'static` because the *type* is not tied to a lifetime since it manages its own memory using `String` instead of borrowing from some other memory that lives…somewhere else. Basically, instead of `Cow<'a, str>` for any `'a` being allowed, we're saying `Cow<'static, str>` which is either a `&'static str` or an owned string. Both know when they need deallocated.

Reply:
My understanding is that an object having lifetime `'a` means that it lives at least as long as `'a`. So a lifetime of `'static` implies that the object must live for the lifetime of the program, no? Perhaps I'm missing something...

Reply:
For values, that's true. For *types*, a lifetime is the maximum time the type is valid for. The [`std::Error::source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) method requires a `'static` bound on its return in order to support downcasting. Since lifetimes don't exist at runtime, the only way you can validly downcast is if the type is not tied to any specific lifetime itself because there's no way to know that some lifetime-bound type is actually living long enough to support the cast.